opentracing/src/main/java/opentracing/SpanPropagator.java

```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package opentracing;
+
+import java.util.Map;
+
+/**
+ * Encodes or Decodes a {@linkplain Span} in binary or text formats.
+ *
+ * <p>The toXXX methods are expected to serialize spans into a pair of values representing
+ * separately the span / span identity, and the trace attributes. This is done specifically
+ * for binary protocols that may represent tracing identity in a dedicated fixed-length slot in the
+ * binary message format, so that it can be inspected efficiently by the middleware / routing layers
+ * without parsing the whole message.
+ */
+public interface SpanPropagator {
+
+  /**
+   * Encodes the span into a binary representation of the span's identity and trace
+   * attributes.
+   */
+  byte[] toBinary(Span span);
```

Comment:
Ok. Then we have different notions as to who will implement the IPC components.

I've always been under the impression that this is primarily the concerns of the implementors. For example Zipkin will provide OpenTracing instrumentation libraries to work with http and with thrift and what else.

I still struggle to see why the application developer (even those at the lower level that deal with the IPC code) need to be playing around with the internals of the SpanPropagator, and in the situation where they are why we would not expect them to be doing so using the internal API.

I'm trying to get my head around it, but as far as I can see, adding attributes or bytes to the wire within the SpanPropagator is beyond the concerns of instrumentation: beyond the concerns of the end-user; and moving into more nuts and bolts and hence more the internal API. I wouldn't be surprised either if this type of development is stuff: for example proxy routing where trace attributes are added; we'd be hoping to see contributed back to the particular OpenTracing implementation.